File: src/main/java/com/badbones69/crazycrates/support/structures/StructureHandler.java
package com.badbones69.crazycrates.support.structures;

import com.badbones69.crazycrates.api.CrazyManager;
import com.badbones69.crazycrates.support.structures.interfaces.StructureControl;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.structure.Mirror;
import org.bukkit.block.structure.StructureRotation;
import org.bukkit.structure.StructureManager;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class StructureHandler implements StructureControl {


    private final File file;

    public StructureHandler(File file) {
        this.file = file;
    }

    private final CrazyManager crazyManager = CrazyManager.getInstance();

    private final ArrayList<Block> structureBlocks = new ArrayList<>();

    private final ArrayList<Block> preStructureBlocks = new ArrayList<>();

    StructureManager structureManager = null;

    @Override
    public void pasteStructure(Location location) {
        try {

            structureManager = crazyManager.getPlugin().getServer().getStructureManager();

            getNearbyBlocks(location);

            structureManager.loadStructure(file).place(location, false, StructureRotation.NONE, Mirror.NONE, 0, 1F, new Random());

            getStructureBlocks(location);
        } catch (Exception e) {
            crazyManager.getPlugin().getLogger().warning(e.getMessage());
        }
    }

    @Override
    public void removeStructure(Location location) {
        structureBlocks.forEach(block -> block.getLocation().getBlock().setType(Material.AIR));
    }

    @Override
    public void saveSchematic(Location[] locations) {
        // Not available for use.
    }

    @Override
    public double getStructureX() {
        try {
            return structureManager.loadStructure(file).getSize().getX();
        } catch (Exception e) {
            crazyManager.getPlugin().getLogger().warning(e.getMessage());
        }

        return 0;
    }

    @Override
    public double getStructureZ() {
        try {
            return structureManager.loadStructure(file).getSize().getZ();
        } catch (Exception e) {
            crazyManager.getPlugin().getLogger().warning(e.getMessage());
        }

        return 0;
    }

    @Override
    public List<Material> getBlackList() {
        List<Material> blockList = new ArrayList<>();

        blockList.add(Material.ACACIA_SIGN);
        blockList.add(Material.BIRCH_SIGN);
        blockList.add(Material.DARK_OAK_SIGN);
        blockList.add(Material.JUNGLE_SIGN);
        blockList.add(Material.OAK_SIGN);
        blockList.add(Material.SPRUCE_SIGN);
        blockList.add(Material.ACACIA_WALL_SIGN);
        blockList.add(Material.BIRCH_WALL_SIGN);
        blockList.add(Material.DARK_OAK_WALL_SIGN);
        blockList.add(Material.JUNGLE_WALL_SIGN);
        blockList.add(Material.OAK_WALL_SIGN);
        blockList.add(Material.SPRUCE_WALL_SIGN);
        blockList.add(Material.STONE_BUTTON);
        blockList.add(Material.BIRCH_BUTTON);
        blockList.add(Material.ACACIA_BUTTON);
        blockList.add(Material.DARK_OAK_BUTTON);
        blockList.add(Material.JUNGLE_BUTTON);
        blockList.add(Material.OAK_BUTTON);
        blockList.add(Material.SPRUCE_BUTTON);
        blockList.add(Material.STONE_BUTTON);

        return blockList;
    }

    @Override
    public List<Block> getStructureBlocks(Location location) throws IOException {

        for (int x = 0; x < structureManager.loadStructure(file).getSize().getX(); x++) {
            for (int y = 0; y < structureManager.loadStructure(file).getSize().getY(); y++) {
                for (int z = 0; z < structureManager.loadStructure(file).getSize().getZ(); z++) {
                    structureBlocks.add(location.getBlock().getRelative(x, y, z));
                }
            }
        }

        return structureBlocks;
    }

    @Override
    public List<Block> getNearbyBlocks(Location location) throws IOException {

        for (int x = 0; x < structureManager.loadStructure(file).getSize().getX(); x++) {
            for (int y = 0; y < structureManager.loadStructure(file).getSize().getY(); y++) {
                for (int z = 0; z < structureManager.loadStructure(file).getSize().getZ(); z++) {
                    preStructureBlocks.add(location.getBlock().getRelative(x, y, z));
                }
            }
        }

        return preStructureBlocks;
    }
}

File: src/main/java/com/badbones69/crazycrates/support/structures/StructureHandler.kt
package com.badbones69.crazycrates.support.structures

import com.badbones69.crazycrates.support.structures.interfaces.StructureControl
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.structure.Mirror
import org.bukkit.block.structure.StructureRotation
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.util.*

class StructureHandler(private val plugin: JavaPlugin, val file: File) : StructureControl {

    private val structureManager = plugin.server.structureManager.loadStructure(file)

    private val structureBlocks = arrayListOf<Block>()

    private val preStructureBlocks = arrayListOf<Block>()

    override fun pasteStructure(location: Location) {
        runCatching {
            // Save the old blocks
            getNearbyBlocks(location)

            structureManager.place(location, false, StructureRotation.NONE, Mirror.NONE, 0, 1F, Random())
        }.onFailure { plugin.logger.warning(it.message) }.onSuccess {
            // Save the structure blocks
            getStructureBlocks(location)
        }
    }

    // Don't need yet
    override fun saveStructure(location: Array<Location?>?) {
        runCatching {

        }.onFailure { plugin.logger.warning(it.message) }
    }

    override fun removeStructure(location: Location) {
        structureBlocks.forEach {
            it.location.block.type = Material.AIR
        }
    }

    override fun getStructureX(): Double {
        return structureManager.size.x
    }

    override fun getStructureZ(): Double {
        return structureManager.size.z
    }

    override fun getStructureBlocks(location: Location): ArrayList<Block> {
        for (x in 0 until structureManager.size.x.toInt()) {
            for (y in 0 until structureManager.size.y.toInt()) {
                for (z in 0 until structureManager.size.z.toInt()) {
                    structureBlocks.add(location.block.getRelative(x, y, z))
                }
            }
        }

        structureBlocks.forEach { it.location.block.state.update() }

        return structureBlocks
    }

    override fun getNearbyBlocks(location: Location): ArrayList<Block> {
        for (x in 0 until structureManager.size.x.toInt()) {
            for (y in 0 until structureManager.size.y.toInt()) {
                for (z in 0 until structureManager.size.z.toInt()) {
                    preStructureBlocks.add(location.block.getRelative(x, y, z))
                }
            }
        }
        return preStructureBlocks
    }

    override fun getBlackList(): List<Material> {
        return arrayListOf(
            Material.ACACIA_BUTTON,
            Material.BIRCH_BUTTON,
            Material.DARK_OAK_BUTTON,
            Material.JUNGLE_BUTTON,
            Material.OAK_BUTTON,
            Material.SPRUCE_BUTTON,
            Material.STONE_BUTTON,
            // Add all signs
            Material.ACACIA_SIGN,
            Material.BIRCH_SIGN,
            Material.DARK_OAK_SIGN,
            Material.JUNGLE_SIGN,
            Material.OAK_SIGN,
            Material.SPRUCE_SIGN,
            Material.ACACIA_WALL_SIGN,
            Material.BIRCH_WALL_SIGN,
            Material.DARK_OAK_WALL_SIGN,
            Material.JUNGLE_WALL_SIGN,
            Material.OAK_WALL_SIGN,
            Material.SPRUCE_WALL_SIGN
        )
    }
}